zerver/tests/test_link_embed.py

```diff
@@ -374,3 +375,116 @@ def test_link_embed_data_from_cache(self) -> None:
             key = preview_url_cache_key(url)
             cache_set(key, link_embed_data, 'database')
             self.assertEqual(link_embed_data, link_embed_data_from_cache(url))
+
+    @override_settings(INLINE_URL_EMBED_PREVIEW=True)
+    def test_link_preview_non_html_data(self) -> None:
+        email = self.example_email('hamlet')
+        self.login(email)
+        url = 'http://test.org/audio.mp3'
+        with mock.patch('zerver.lib.actions.queue_json_publish') as patched:
+            msg_id = self.send_stream_message(email, "Scotland", topic_name="foo", content=url)
+            patched.assert_called_once()
+            queue = patched.call_args[0][0]
+            self.assertEqual(queue, "embed_links")
+            event = patched.call_args[0][1]
+
+        headers = {'content-type': 'application/octet-stream'}
+        mocked_response = mock.Mock(side_effect=self.create_mock_response(url, headers=headers))
+
+        with self.settings(TEST_SUITE=False, CACHES=TEST_CACHES):
+            with mock.patch('requests.get', mocked_response):
+                FetchLinksEmbedData().consume(event)
+
+                cached_data = link_embed_data_from_cache(url)
```

Comment:
Just a note on test style for this system: Long after this test suite was originally written, we changed the default behavior of `queue_json_publish` in the test suite to just call the processor function (it reaches into `zerver/worker/queue_processors.py` to figure out which that is), which provides much better test coverage and requires a lot less messy locking.  So aside from the couple tests in this file where we're expressly trying to test the situation before the queue processor runs, we should be able to simplify these tests a lot.

Secondly, for mocking HTTP requests, there's an option to use the standard `httpretty` library rather than mocking `requests.get`; might not be right for some use cases, but I wanted to mention it since it can be handy (we use it in our authentication backend tests).

Reply:
Yeah, the test suite did look more complicated than other places that I touched in Zulip, but didn't investigate why. I'll simplify the tests, at some point.